README.md
# What's this?

A CHIP8 Interpreter written with OpenGL

### Progress

  * All CHIP-8 opcodes are implemented
  * Input mostly function in OpenGL version, bit laggy though
  * Display and keyboard support implemented with OpenGL and GLFW.
  * Interpreter can run at selected clock cycles between 1 and 100 per frame. (Adjust accordingly to your display refresh rate)
  * CHIP-8, HiRES and Super-CHIP8 support.


Changelog: https://github.com/Skryptonyte/CHIP8-Interpreter/blob/opengl/CHANGELOG.md
### What works?

All CHIP-8 games and SUPER CHIP-8 games ive tested so far work

### TODO
  * Add enhancements for CHIP-8
    * Mega CHIP-8 Instruction set
    * Add some tweaks for certains games with different behaviour
  * Option to disassemble to a file
  

### Requirements for compilation and use

  * GCC/Clang ( or MinGW on Windows )
  * OpenGL, GLFW library


To compile: 
`
gcc driver.c -lglfw -lpthread -lGLEW -lGL 
`


CHANGELOG.md


# Release 1.2.0
  * Scroll functions implemented properly (except half scroll), making SCHIP-8 support 99% complete
  * Input system implemented for OpenGL version, Q W E R A S D F Z X C V will act as the emulated keypad.
  * Option to adjust cycles per frame
  

# Release 1.1.0

  * RND function now seeded at the start of execution to make it more random
  * Vertical scrolling implemented ( I don't even know whether this works ). Half pixel scroll (Scrolling in non-extended mdoe) doesn't and probably never will due to the limitations of the implementation of my display. I'd need to do a complete rewrite of the display to get 64x32 to scale to a 128x64 display size.
  * Massive code clean up
  * 10 byte fontset added
  * HiRes mode initialization fixed.

  * New OpenGL branch. (EXPERIMENTAL and UNOPTIMIZED!)

# Release 1.0.1a
  * DRW Instruction had a bug where sprites would appear to have vertical scanlines. This was causes by forgetting to draw the 8th bit of each byte of the sprite.
  * DRW instruction's SCHIP extension turned out to be severely broken. I've fixed that.

# Release 1.0.1

  * RPL User flags added
  * New instructions implemented:
    * 0xFE: Disable Extended mode (128x64)
    * 0xFF: Enable Extended mode
    * Fx75 and Fx85: Move data from registers to RPL flags and vice versa
  * DRW opcode extended: DXY0 will draw 16x16 sprites
  * Delay/Sound Timer now runs at the correct clock-rate. The timers will update exactly every 16.667ms on a separate thread

# Release 1.0.0

  * Display of all the CHIP 8 registers below the display output
  * Instruction Fx18 implemented. Sound Timer however is effectively a stub and behaves just like the Delay Timer for the time being until I figure out low level audio programming.
  * HiRes mode added (64x64 display)
  * Preliminary Super Chip-8 addition: 128x64 display mode


driver.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include "pthread.h"

#include <time.h>
#include "chip8core.h"
#include <signal.h>




void windcall(GLFWwindow* win, int width, int height){
	glViewport(0,0,width,height);
	puts("Window size changing");
}

void key_callback(GLFWwindow* window, int key, int scancode, int action, int mods)
{
    char key_table[16] = {'X','1','2','3',
			'Q','W','E',
			'A','S','D',
			'Z','C','4','R','F','V'};
    if (action == GLFW_PRESS){
		for (int i = 0; i < 16; i++)
		{

			if (key == key_table[i])
			{
				keyb = i;
				break;
			}
		}

	}

   if (action == GLFW_RELEASE)
   {
	keyb = 0xfe;
   }
}


void* delayHandler(void* meh){
while (1){
	if (dt > 0){
		dt--;
	}
	if (st > 0){
		st--;
	}
	usleep(16666);
	}
	return NULL;
}

void handler(int sig){

puts("Invalid Memory Access..");
printf("Offending OpCode: %x\nProgram Counter: %x\n",current_opcode, pc);
puts("Registers: ");
for (int i = 0; i <= 0xf; i++){
printf("V%x: %d\n",i,v[i]);
}
printf("I: %x, PC = %x\n", i, pc);
exit(sig);
}
void initialize(){
//puts("Initializing system..");
memset(v,0, 17*sizeof(unsigned char));
pc = 0x200;
sp = 0;
current_opcode = 0;
i = 0;
dpflag = 0;
keyb = 0;

dt = 0;
st = 0;
}

void loadGame(char* c){
FILE* fp = fopen(c,"rb");

if (!fp){

	puts("Game not found... exitting");
	exit(1);
	return;
}
unsigned char ch = getc(fp);
int size = 0;
for (int i = 512; i < 4096; i++){
	memory[i] = ch;
	size += 1;
	ch = getc(fp);
}
}

void emulateCycle(int cycle_count){
	
	for (int i = 1; i <= cycle_count; i++){
	glfwPollEvents();
	unsigned short fetch = (memory[pc] << 8) | memory[pc+1];
	decodeOp(fetch);
	pc += 2;
	}
	
}

void displayScreen(int VBO, int prog){
float xoffset = 2.0/((float) display_x);
float yoffset = 2.0/((float) display_y);

float vertices[] = {-1.0f,1.0f,
		    -1.0f+xoffset, 1.0f,
		    -1.0f, 1.0f-yoffset,
		     -1.0f+xoffset, 1.0f-yoffset};

int y = 0;
while (y < display_y){
int x = 0;
while (x < display_x){
if (display[x][y] == 1){
glBufferData(GL_ARRAY_BUFFER,sizeof(vertices), vertices, GL_STREAM_DRAW);
glDrawArrays(GL_TRIANGLE_STRIP,0,4);
}
int i = 0;
while (i < 8){
vertices[i] += xoffset;
i += 2;
}
x++;
}

int i = 1;
while (i < 8){
vertices[i] = vertices[i] - yoffset;
i += 2;
}

vertices[0] = -1.0;
vertices[2] = -1.0+xoffset;
vertices[4] = -1.0;
vertices[6] = -1.0+xoffset;
y++;
}
}
void getSource(char* filename, int shader){
	FILE* f = fopen(filename,"r");
	if (!f){
		printf("Missing file: %s",filename);
		exit(-1);
	}
	char buffer[512];
	char ch = getc(f);
	int index = 0;
	while (ch != EOF){
		buffer[index] = ch;
		ch = getc(f);
		index++;
	}
	buffer[index] = 0;
	const char* data = buffer;
	glShaderSource(shader,1,&data,NULL);
}

void compileSource(int shader){
glCompileShader(shader);

int successStatus, shaderType;
glGetShaderiv(shader,GL_COMPILE_STATUS,&successStatus);
glGetShaderiv(shader,GL_SHADER_TYPE,&shaderType);

if (successStatus == GL_FALSE){
	puts("Shader Compilation failed..");
	exit(-1);
}
}
int main(int argc, char *argv[]){

signal(SIGSEGV, handler);

initialize();
fillFont();
int clock_cycles = 10;
if (argc < 2){
	puts("Syntax: ./programfile <chip8 rom> <clocks per frame (default: 10)");
	exit(1);
	}

if (argc == 3)
{
	int a = strtol(argv[2],NULL,10);
	if (a >= 1 && a <= 100)
	{
		clock_cycles = a;
	}
	else{
		puts("Invalid cycle count, defaulting to 10)");
	}
}


puts("Initializing emulated CPU");

srand(time(0));

loadGame(argv[1]);


pthread_t pid, pid2;
pthread_create(&pid,NULL,delayHandler,NULL);

int cycle = 1;
if ((memory[pc] << 8 | memory[pc+1]) == 0x1260){
display_x = 64;
display_y = 64;
pc = 0x2c0 - 2;
}

puts("Initializing OpenGL Renderer");

glfwInit();
glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR,4);
glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR,6);

GLFWwindow* win = glfwCreateWindow(1920,1080,"CHIP8 Emulator",NULL, NULL);
glfwMakeContextCurrent(win);

glewInit();

glfwSetWindowSizeCallback(win, windcall);
glfwSetKeyCallback(win,key_callback);

int vertexShader = glCreateShader(GL_VERTEX_SHADER);
int fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);

getSource("vertex.glsl",vertexShader);
getSource("fragment.glsl",fragmentShader);

compileSource(vertexShader);
compileSource(fragmentShader);

int prog = glCreateProgram();
glAttachShader(prog,vertexShader);
glAttachShader(prog,fragmentShader);

glLinkProgram(prog);
int linkStatus;
glGetProgramiv(prog,GL_LINK_STATUS, &linkStatus);

if (linkStatus == GL_FALSE){
	puts("Linking failed..");
	char data[512];
	glGetProgramInfoLog(prog,512,NULL,data);
	puts("INFO LOG: ");
	printf("%s\n",data);
	return -1;
}
puts("Shaders compiled and linked");

int VBO;
glGenBuffers(1,&VBO);
glBindBuffer(GL_ARRAY_BUFFER,VBO);

int VAO;
glGenVertexArrays(1,&VAO);
glBindVertexArray(VAO);
glEnableVertexAttribArray(0);
glVertexAttribPointer(0,2,GL_FLOAT,GL_FALSE, sizeof(float) *2, (void*)0);

glUseProgram(prog);
glfwSetWindowSizeCallback(win, windcall);

while (1){
emulateCycle(clock_cycles);

glClearColor(0.1,0.1,0.1,1.0);
glClear(GL_COLOR_BUFFER_BIT);
//if (dpflag){
displayScreen(VBO,prog);
glfwSwapBuffers(win);
//}


}
//usleep(2000);
}


chip8core.h
#include <GL/glew.h>
#include <GLFW/glfw3.h>

#include <stdlib.h>
#include <string.h>
#include <math.h>

unsigned char keyb = 0xfe;
unsigned char lastkey;
unsigned char key = 255;

int current_opcode;

unsigned char dpflag = 0;

unsigned char v[17];
unsigned char rpl[8];

unsigned short i;
unsigned short pc;

short int display_x = 64;
short int display_y = 32;

int y_wrap = 1;
unsigned char fontset[0x5 * 16] = {0xf0, 0x90, 0x90, 0x90, 0xf0,   //0
				0x20, 0x60, 0x20, 0x20, 0x70,	//1
				0xf0, 0x10,0xf0, 0x80, 0xf0,	//2
				0xf0, 0x10,0xf0, 0x10, 0xf0	//3
				,0x90, 0x90, 0xf0, 0x10, 0x10	//4
				,0xf0, 0x80, 0xf0, 0x10, 0xf0	//5
				,0xf0, 0x80, 0xf0, 0x90, 0xf0	//6
				,0xf0, 0x10, 0x20, 0x40, 0x40   //7
				,0xf0, 0x90, 0xf0, 0x90, 0xf0	//8
				,0xf0, 0x90, 0xf0, 0x10, 0xf0	//9
				,0xf0, 0x90, 0xf0, 0x90, 0x90,	//A
				0xe0, 0x90, 0xe0, 0x90, 0xe0,	//B
				0xf0, 0x80, 0x80, 0x80, 0xf0,	//C
				0xe0, 0x90, 0x90, 0x90, 0xe0,	//D
				0xf0, 0x80, 0xf0, 0x80, 0xf0,	//E
				0xf0, 0x80, 0xf0, 0x80, 0x80}	//F
				;
unsigned char hfontset[0x10 * 10] = {
0x7C, 0x82, 0x82, 0x82, 0x82, 0x82, 0x82, 0x82, 0x7C, 0x00,
0x08, 0x18, 0x38, 0x08, 0x08, 0x08, 0x08, 0x08, 0x3C, 0x00,
0x7C, 0x82, 0x02, 0x02, 0x04, 0x18, 0x20, 0x40, 0xFE, 0x00,
0x7C, 0x82, 0x02, 0x02, 0x3C, 0x02, 0x02, 0x82, 0x7C, 0x00,
0x84, 0x84, 0x84, 0x84, 0xFE, 0x04, 0x04, 0x04, 0x04, 0x00,
0xFE, 0x80, 0x80, 0x80, 0xFC, 0x02, 0x02, 0x82, 0x7C, 0x00,
0x7C, 0x82, 0x80, 0x80, 0xFC, 0x82, 0x82, 0x82, 0x7C, 0x00,
0xFE, 0x02, 0x04, 0x08, 0x10, 0x20, 0x20, 0x20, 0x20, 0x00,
0x7C, 0x82, 0x82, 0x82, 0x7C, 0x82, 0x82, 0x82, 0x7C, 0x00,
0x7C, 0x82, 0x82, 0x82, 0x7E, 0x02, 0x02, 0x82, 0x7C, 0x00,
0x10, 0x28, 0x44, 0x82, 0x82, 0xFE, 0x82, 0x82, 0x82, 0x00,
0xFC, 0x82, 0x82, 0x82, 0xFC, 0x82, 0x82, 0x82, 0xFC, 0x00,
0x7C, 0x82, 0x80, 0x80, 0x80, 0x80, 0x80, 0x82, 0x7C, 0x00,
0xFC, 0x82, 0x82, 0x82, 0x82, 0x82, 0x82, 0x82, 0xFC, 0x00,
0xFE, 0x80, 0x80, 0x80, 0xF8, 0x80, 0x80, 0x80, 0xFE, 0x00,
0xFE, 0x80, 0x80, 0x80, 0xF8, 0x80, 0x80, 0x80, 0x80, 0x00
};
unsigned char memory[4096];

unsigned short stack[16];
unsigned short sp;

wchar_t states[2] = {' ',0x2588};
unsigned char display[128][64];

unsigned char dt;
unsigned char st;
void fillFont(){
for (int i = 0; i < 0x200; i++){
memory[i] = fontset[i];
}


for (int i = 0; i < 0x200; i++){
memory[i + 0x5*16] = hfontset[i];
}
}
void decodeOp(unsigned short fetchedOpcode){
current_opcode = fetchedOpcode;

unsigned short opcode = fetchedOpcode >> 12;

unsigned short fh = fetchedOpcode >> 8;
unsigned short sh = fetchedOpcode << 8;
sh >>= 8;
unsigned char firsthalf = fh;
unsigned char secondhalf = sh;

unsigned char secbyte = (firsthalf << 4);
secbyte >>= 4;

unsigned char seclastbyte = secondhalf >> 4;
unsigned char lastbyte = secondhalf << 4;
lastbyte >>= 4;
unsigned short lastthreebyte = (fetchedOpcode << 4);
lastthreebyte >>= 4;

switch (opcode){
	case 0x0:
		switch (seclastbyte){
		case 0xc:
		for (int j = display_y-1; j >= lastbyte; j--)
		{
			for (int i = 0; i < display_x; i++)
			{
				display[i][j] = display[i][j-lastbyte];

			}
		}

		for (int j = 0; j < lastbyte; j++)
		{
			for (int i =0; i < display_x; i++)
			{
				display[i][j] = 0;
			}
		}

			return;
		}
		{
		switch (secondhalf){
			case 0xe0:
				for (int i = 0; i < display_x; i++){
					for (int j = 0; j < display_y; j++){
						display[i][j] = 0;
					}
				}
				dpflag ^= 1;
				return;
			case 0xee:
				pc = stack[sp-1];
				sp--;
				return;
			case 0xfb:

					for (int i = display_x-1;i >= 4; i--){
						for (int j = 0; j < display_y; j++){
						display[i][j] = display[i-4][j];
						}
					}

					for (i = 0; i < 4; i++){
						for (int j = 0; j < display_y; j++)
						{
							display[i][j] = 0;
						}
					}
					return;
			case 0xfc:
					for (int i = 0; i < display_x-4; i++)
					{
						for (int j = 0; j < display_y; j++)
						{
							display[i][j] = display[i+4][j];
						}
					}


					for (int i =display_x-4-1; i < display_x;i++)
					{
						for (int j = 0; j < display_y; j++)
						{
							display[i][j] = 0;
						}
					}
					return;
			case 0xfd:

				puts("Exit requested by program..");
				exit(0);
				return;
			case 0xfe:
				display_x = 64;
				display_y = 32;
				return;
			case 0xff:
				display_x = 128;
				display_y = 64;
				return;
			default:
				return;
		}
		}
	case 0x1:
		pc = lastthreebyte - 0x2;
		return;
	case 0x2:
		sp++;
		stack[sp-1] = pc;
		pc = lastthreebyte - 0x2;
		return;
	case 0x3:
		if (v[secbyte] == secondhalf){
			pc += 2;
		}
		return;
	case 0x4:
		if ((v[secbyte] ^ secondhalf)){
			pc += 2;
		}
		return;
	case 0x5:
		if (lastbyte == 0){
		if (v[secbyte] == v[seclastbyte]){
			pc += 2;
		}
		}
		return;
	case 0x6:
		v[secbyte] = secondhalf;
		return;
	case 0x7:
		v[secbyte] += secondhalf;
		return;
	case 0x8:
		switch (lastbyte){
			case 0x0:
				v[secbyte] = v[seclastbyte];
				return;
			case 0x1:
				v[secbyte] |= v[seclastbyte];
				return;
			case 0x2:
				v[secbyte] &= v[seclastbyte];
				return;
			case 0x3:
				v[secbyte] ^= v[seclastbyte];
				return;
			case 0x4:
				if (v[secbyte] + v[seclastbyte] > 255){
					v[0xf] = 1;
				}
				v[secbyte] += v[seclastbyte];
				return;
			case 0x5:
				if (v[seclastbyte] > v[secbyte]){
					v[0xf] = 0x0;
				}
				else {
					v[0xf] = 0x1;
				}
				v[secbyte] -= v[seclastbyte];
				return;
			case 0x6:
				v[0xf] = v[secbyte] & 1;
				v[secbyte] >>= 1;
				return;
			case 0x7:
				if (v[secbyte] > v[seclastbyte]){
					v[0xf] = 0x0;
				}
				else{
					v[0xf] = 0x1;
				}
				v[secbyte] = v[seclastbyte] - v[secbyte];
				return;
			case 0xe:
				v[0xf] = v[secbyte] >> 7;
				v[secbyte] = v[secbyte] << 1;
				return;
				}
	case 0x9:
		if (v[secbyte] != v[seclastbyte]){
			pc += 2;
		}
		return;
	case 0xa:
		i = lastthreebyte;
		return;
	case 0xb:
		pc = v[0] + lastthreebyte - 0x2;
		return;
	case 0xc:
		v[secbyte] = rand() % 0xff;
		v[secbyte] &= secondhalf;
		return;
	case 0xd:
		v[0xf] = 0x0;
		unsigned char bitcount = 7;
		unsigned char inc = 1;
		if (lastbyte == 0){
			bitcount = 15;
			lastbyte = 16;
		}
		int ycount = 0;
		for (unsigned char y = 0; y < lastbyte; y++){
			unsigned short sprite = (memory[i +ycount] << 8)|memory[i+ycount+1];
			int count = 0;
			for (char x = bitcount; x >=0; x--){
				int y_coordinate = v[seclastbyte] + y;
				if (y_wrap)
				{
				   y_coordinate = y_coordinate%display_y;
			  }
				else{
						y_coordinate = y_coordinate <0 ? (y_coordinate >= display_y?display_y-1:y_coordinate):0;
				}
				unsigned char prev = display[((v[secbyte])+count)%display_x][y_coordinate];
				unsigned char* ptr = &display[((v[secbyte])+count)%display_x][y_coordinate];
				if (bitcount == 7){
				display[(v[secbyte]+count)%display_x][y_coordinate] ^= ((memory[i+y]) >> (x)) & 1;
				}
				else if (bitcount == 15){
				unsigned short b = (sprite >> x) & 1;
				display[(v[secbyte]+count)%display_x][y_coordinate] ^= b;
				}
				count++;
				if (prev != *ptr && prev == 1){
					v[0xf] = 1;
				}
			}
			ycount += 2;
		}
		dpflag ^= 1;
		return;
	case 0xe:
		switch (secondhalf){
			case 0x9e:
				{
				if (v[secbyte] == keyb){
					pc += 2;
				}
				return;
				}
			case 0xa1:
				{
				if (v[secbyte] != keyb){
					pc += 2;
				}
				return;
				}
	}
	case 0xf:
		switch (secondhalf){
			case 0x07:
				v[secbyte] = dt;
				return;
			case 0xa:

				glfwWaitEvents();

				
				if (keyb >= 0 && keyb <= 15){
					v[secbyte] = keyb;
					break;
				}
				

				return;
			case 0x15:
				dt = v[secbyte];
				return;
			case 0x18:
				st = v[secbyte];
				return;
			case 0x1e:
				if (i + v[secbyte] > 0xFFF){
					v[0xf] = 0x1;
				}
				else{
					v[0xf] = 0x0;
				}
				i += v[secbyte];
				return;
			case 0x29:
				i = (5 * v[secbyte]);
				return;
			case 0x30:
				i = ((16*0x5)+ 10 * v[secbyte]);
				return;
			case 0x33:
				memory[i] = (v[secbyte] / 100);
				memory[i+1] = (v[secbyte] % 100);
				memory[i+1] /= 10;
				memory[i+2] = (v[secbyte] % 10);
				return;
			case 0x55:
				for (int V = 0; V <= secbyte; V++){
				memory[i+V] = v[V];
				}
				return;
			case 0x65:
				for (int V = 0; V <= secbyte; V++){
				v[V] = memory[i + V];
				}
				return;
			case 0x75:
				for (int V = 0; V <= secbyte && V <= 7; V++){
					rpl[V] = v[V];
				}
				return;
			case 0x85:
				for (int V = 0; V <= secbyte && V <= 7; V++){
					v[V] = rpl[V];
				}
				return;
			default:
				return;
			}
		}

}
